lib/profileAgents.js

```diff
@@ -602,6 +609,104 @@ async function _updateMeterController({meterId, controller}) {
   }
 }
 
+async function _getProfileAgentRecords({accountId, includeSecrets}) {
+  const query = {'profileAgent.account': accountId};
+  const projection = {_id: 0};
+  const collection = getCollection('profile-profileAgent');
+  let records = await collection.find(query, {projection}).toArray();
+  const zcapRefreshed = await _refreshProfileAgentZcaps({
+    profileAgentRecords: records
+  });
+  if(zcapRefreshed) {
+    records = await collection.find(query, {projection}).toArray();
+  }
+  if(!includeSecrets) {
+    records = await removeSecretsFromRecords({records});
+  }
+  return records;
+}
+
+async function _refreshProfileAgentZcaps({profileAgentRecords}) {
+  let zcapRefreshed = false;
+  const records = klona(profileAgentRecords);
+  const now = Date.now();
+  for(const record of records) {
+    if(!record.profileAgent.zcaps) {
+      continue;
+    }
+    const {zcaps} = record.profileAgent;
+    for(const zcapName in zcaps) {
+      if(zcapName !== 'profileCapabilityInvocationKey') {
+        const zcap = zcaps[zcapName];
+        const {expires} = zcap;
+        // check if the zcap is close to expiry
+        const expTime = Date.parse(expires);
+        const {zcapRefreshThreshold} = config.profile.profileAgent;
+        if(expTime - now <= zcapRefreshThreshold) {
+          if(zcapName === 'userDocument') {
+            await _refreshProfileAgentUserDocZcaps({zcap, now, record});
+          }
+          zcap.expires = new Date(now + DEFAULT_PROFILE_AGENT_ZCAP_TTL);
+          zcapRefreshed = true;
+        }
+      }
+    }
+    if(zcapRefreshed) {
+      // update the profile agent record
+      record.profileAgent.sequence += 1;
+      await update({profileAgent: record.profileAgent});
+    }
+  }
```

Comment:
1. Cannot refresh when `profileCapabilityInvocationKey` zcaps isn't present (non-root profile agent).
2. Invert conditionals to reduce indented code size for increased readability.
3. Only fetch config option once.
4. After applying these changes -- we still need to modify this further so it will run every record update in parallel instead of serial. I recommend making another helper for that for each individual record and calling them all with `Promise.all()`.
5. Update can fail due to a conflict error -- this should not cause the overall operation to fail, it should loop and try again.

```suggestion
  const {zcapRefreshThreshold} = config.profile.profileAgent;
  for(const record of records) {
    if(!record.profileAgent.zcaps) {
      continue;
    }
    const {zcaps} = record.profileAgent;
    if(!zcaps?.profileCapabilityInvocationKey) {
      // cannot refresh without this zcap
      continue;
    }
    for(const zcapName in zcaps) {
      if(zcapName === 'profileCapabilityInvocationKey') {
        // cannot refresh this zcap
        continue;
      }
      const zcap = zcaps[zcapName];
      const {expires} = zcap;
      // check if the zcap is close to expiry
      const expTime = Date.parse(expires);
      if((expTime - now) <= zcapRefreshThreshold) {
        if(zcapName === 'userDocument') {
          await _refreshProfileAgentUserDocZcaps({zcap, now, record});
        }
        zcap.expires = new Date(now + DEFAULT_PROFILE_AGENT_ZCAP_TTL);
        zcapRefreshed = true;
      }
    }
    if(zcapRefreshed) {
      // update the profile agent record
      record.profileAgent.sequence += 1;
      await update({profileAgent: record.profileAgent});
    }
  }
```

Reply:
Updated.